### PR TITLE
Support for Kubernetes Exporter

### PR DESCRIPTION
**Pull request by MartinWeindel**
**What this PR does / why we need it**:
Support for `k8sExporter` feature of the network-problem-detector (see PR gardener/network-problem-detector#12).
If enabled, the agents are updating the conditions in the node status at the heartbeat period.
The feature is disabled by default. To enable it, this value needs to be added to the controller deployment:

```yaml
providerConfig:
  values:
    networkProblemDetector:
      k8sExporter:
        enabled: true
```

Additionally, the network policy configuration has been fixed for the agents. This is only relevant in case a deny-all network policy is added to the `kube-system` namespace on the shoot.
 
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature operator
Support for `k8sExporter` feature of the network-problem-detector
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -150,3 +149,9 @@ replace (
 	k8s.io/code-generator => k8s.io/code-generator v0.24.3
 	k8s.io/component-base => k8s.io/component-base v0.24.3
 )
+
+// workaround for https://github.com/gardener/hvpa-controller/issues/92, remove once it's fixed
+replace (
+	github.com/gardener/hvpa-controller => github.com/gardener/hvpa-controller v0.4.0
+	github.com/gardener/hvpa-controller/api => github.com/gardener/hvpa-controller/api v0.4.0
+)
```
Comment (ScheererJ) [inline go.mod:157]: Is this really required?
If I remember correctly, this was removed from other extensions when gardener 1.52 was revendored. The networking problem detector already has gardener 1.52.

File: pkg/controller/lifecycle/actuator.go [hunk 2]
```diff
@@ -282,6 +310,55 @@ func (a *actuator) getShootAgentResources(defaultPeriod time.Duration, pingEnabl
 	return shootResources, nil
 }
 
+func buildAgentNetworkPolicy() client.Object {
+	tcp := corev1.ProtocolTCP
+	podGRPCPort := intstr.FromInt(common.PodNetPodGRPCPort)
+	podHttpPort := intstr.FromInt(common.PodNetPodHttpPort)
+	hostGRPCPort := intstr.FromInt(common.HostNetPodGRPCPort)
```
Comment (ScheererJ) [inline pkg/controller/lifecycle/actuator.go:317]: Is there not a port missing? I would have expected `hostHttpPort` to be here as well.

File: pkg/controller/lifecycle/actuator.go [hunk 3]
```diff
@@ -282,6 +310,55 @@ func (a *actuator) getShootAgentResources(defaultPeriod time.Duration, pingEnabl
...
+					Ports: []networkingv1.NetworkPolicyPort{
+						{
+							Protocol: &tcp,
+							Port:     &podGRPCPort,
+						},
+						{
+							Protocol: &tcp,
+							Port:     &hostGRPCPort,
+						},
+					},
```
Comment (ScheererJ) [inline pkg/controller/lifecycle/actuator.go:340]: Shouldn't here all four ports be mentioned?

File: pkg/controller/lifecycle/actuator.go [hunk 4]
```diff
@@ -282,6 +310,55 @@ func (a *actuator) getShootAgentResources(defaultPeriod time.Duration, pingEnabl
...
+					Ports: []networkingv1.NetworkPolicyPort{
+						{
+							Protocol: &tcp,
+							Port:     &podGRPCPort,
+						},
+						{
+							Protocol: &tcp,
+							Port:     &podHttpPort,
+						},
+					},
```
Comment (ScheererJ) [inline pkg/controller/lifecycle/actuator.go:354]: Same as above, i.e. I would have expected four ports.

File: pkg/controller/lifecycle/actuator.go [hunk 5]
```diff
@@ -282,6 +310,55 @@ func (a *actuator) getShootAgentResources(defaultPeriod time.Duration, pingEnabl
...
+			Name:      "gardener.cloud--allow-to-from-nwpd-agents",
+			Namespace: constants.NamespaceKubeSystem,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"networking.gardener.cloud/to-from-nwpd-agents": "allowed",
+				},
+			},
+			Egress: []networkingv1.NetworkPolicyEgressRule{
```
Comment (ScheererJ) [inline pkg/controller/lifecycle/actuator.go:329]: Can we add a `podSelector` here?

File: pkg/controller/lifecycle/actuator.go [hunk 6]
```diff
@@ -282,6 +310,55 @@ func (a *actuator) getShootAgentResources(defaultPeriod time.Duration, pingEnabl
...
+							Port:     &podGRPCPort,
+						},
+						{
+							Protocol: &tcp,
+							Port:     &hostGRPCPort,
+						},
+					},
+				},
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{
```
Comment (ScheererJ) [inline pkg/controller/lifecycle/actuator.go:343]: Can we add a `podSelector` here?

**Review by ScheererJ — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (MartinWeindel) [reply go.mod:157]: Good catch. This was reintroduced by the merge after rebasing.

Comment (MartinWeindel) [reply go.mod]: fixed

File: pkg/controller/lifecycle/actuator.go (hunk 2 above)
Comment (MartinWeindel) [reply pkg/controller/lifecycle/actuator.go:317]: added hostHttpPort

File: pkg/controller/lifecycle/actuator.go (hunk 4 above)
Comment (MartinWeindel) [reply pkg/controller/lifecycle/actuator.go]: added all four ports for ingress

File: pkg/controller/lifecycle/actuator.go (hunk 3 above)
Comment (MartinWeindel) [reply pkg/controller/lifecycle/actuator.go]: the HTTP ports are only used for exposing metrics, no need to access them from other agents

File: pkg/controller/lifecycle/actuator.go (hunk 5 above)
Comment (MartinWeindel) [reply pkg/controller/lifecycle/actuator.go:330]: added podSelector for the GRPC port on the pod network. For the GRPC port on node network it does not work, probably you would need the IP block of the nodes.

File: pkg/controller/lifecycle/actuator.go (hunk 6 above)
Comment (MartinWeindel) [reply pkg/controller/lifecycle/actuator.go:355]: for ingress a `podSelector` makes no sense, as these ports are also accessed by Prometheus and the `nwpdcli`.

**Pull request merged**